DataStructures/Trees/RedBlackTree.cs

```diff
@@ -455,80 +454,104 @@ protected override bool _remove(BSTNode<TKey> nodeToDelete)
         }
 
         /// <summary>
-        /// The internal remove helper.
-        /// Separated from the overriden version to avoid casting the objects from BSTNode to RedBlackTreeNode.
-        /// This is called from the overriden _remove(BSTNode nodeToDelete) helper.
+        ///     The internal remove helper.
+        ///     Separated from the overriden version to avoid casting the objects from BSTNode to RedBlackTreeNode.
+        ///     This is called from the overriden _remove(BSTNode nodeToDelete) helper.
         /// </summary>
         protected bool _remove(RedBlackTreeNode<TKey> nodeToDelete)
         {
             if (nodeToDelete == null)
-                return false;
-
-            // Temporary nodes
-            RedBlackTreeNode<TKey> node1, node2;
-
-            // If nodeToDelete has either one child or no children at all
-            if (!nodeToDelete.HasLeftChild || !nodeToDelete.HasRightChild)
-            {
-                node1 = nodeToDelete;
-            }
-            else
             {
-                // nodeToDelete has two children
-                node1 = (RedBlackTreeNode<TKey>)base._findNextLarger(nodeToDelete);
+                return false;
             }
 
-            // Left child case
-            if (node1.HasLeftChild)
+            if (!nodeToDelete.HasChildren)
             {
-                node2 = node1.LeftChild;
+                Root = null;
```

Comment:
Why do you nullify the Root of the tree if the node we want to delete doesn't have any children? Can you please add a comment explaining the reason behind this?

Reply:
Yes i will fix that, i don't remember why i done that...

Reply:
Fixed.
I added condition to check that element is root before assign, and created new unit test that cover it.
In line 480 is covered case when we remove node that haven't children and is't root. 

Thanks for your attention, I don't know how I missed such an obvious case.